### ISSUE TITLE
[JENKINS-73239] Active Choice plugin does not handle parameters as expected

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/active-choices-plugin#866]
<p>We have long used the active choice plugin to render our own Build with Parameters page.</p>

<p>It has needed a user script to work properly for a long time, but after recent updates it does not work with that either.</p>

<p>Expected: The rendered page shows default values. When changing from default values, the values are updated.</p>

<p>Actual: The rendered page shows empty checkboxes, text fields, etc. When changing parameters, the default parameters are still used during execution.</p>

<p>Attachments: Sample Jenkinsfile, log, screenshots of sample UI and actual UI.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gtv">gtv</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239" target="_blank">Active Choice plugin does not handle parameters as expected</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow">kinow</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: active-choices-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2025-04-06T19:40:33+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-12-08
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We have long used the active choice plugin to render our own Build with Parameters page.</p>

<p>It has needed a user script to work properly for a long time, but after recent updates it does not work with that either.</p>

<p>Expected: The rendered page shows default values. When changing from default values, the values are updated.</p>

<p>Actual: The rendered page shows empty checkboxes, text fields, etc. When changing parameters, the default parameters are still used during execution.</p>

<p>Attachments: Sample Jenkinsfile, log, screenshots of sample UI and actual UI.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Active Choices Plug-in 2.8.3<br/>
Jenkins 2.452.1
```
</details>
<details><summary><i>7 attachments</i></summary>

- [Jenkinsfile.sample](https://issues.jenkins.io/secure/attachment/62653/Jenkinsfile.sample)
- [UI_actual.jpg](https://issues.jenkins.io/secure/attachment/62652/UI_actual.jpg)
  > ![UI_actual.jpg](https://issues.jenkins.io/secure/attachment/62652/UI_actual.jpg)
- [UI_sample.jpg](https://issues.jenkins.io/secure/attachment/62651/UI_sample.jpg)
  > ![UI_sample.jpg](https://issues.jenkins.io/secure/attachment/62651/UI_sample.jpg)
- [image-2024-07-06-10-43-20-297.png](https://issues.jenkins.io/secure/attachment/62948/image-2024-07-06-10-43-20-297.png)
  > ![image-2024-07-06-10-43-20-297.png](https://issues.jenkins.io/secure/attachment/62948/image-2024-07-06-10-43-20-297.png)
- [image-2025-04-06-19-38-17-470.png](https://issues.jenkins.io/secure/attachment/64211/image-2025-04-06-19-38-17-470.png)
  > ![image-2025-04-06-19-38-17-470.png](https://issues.jenkins.io/secure/attachment/64211/image-2025-04-06-19-38-17-470.png)
- [image-2025-04-06-19-40-19-325.png](https://issues.jenkins.io/secure/attachment/64212/image-2025-04-06-19-40-19-325.png)
  > ![image-2025-04-06-19-40-19-325.png](https://issues.jenkins.io/secure/attachment/64212/image-2025-04-06-19-40-19-325.png)
- [log.txt](https://issues.jenkins.io/secure/attachment/62650/log.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=448157&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-448157">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi, thanks for the detailed report. I created a new pipeline with your example, then played a bit with the returned Groovy (see screenshot).</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/62948/62948_image-2024-07-06-10-43-20-297.png" style="border: 0px solid black" /></span></p>

<p>And I believe what's happening here is that <a href="https://issues.jenkins.io/browse/JENKINS-71909?filter=14783" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-71909?filter=14783</a> is affecting your script now too.</p>

<p>Previously, the parameters were loaded in order. So your document.getElementById would succeed as the previous parameter ("I", with the "CB_BUILDS") would have been rendered first.</p>

<p>This issue linked above was a tentative to optimize the rendering of parameters, by rendering them as fast as possible. Which at least sounds like a good idea, but seems like it's creating issues for several users.</p>

<p>I will not work on your issue for now, and instead will try to use this remaining spare time that I have today/tomorrow to revert that change.</p>

<p>Please, note that if this takes a bit, it's because I work on the plug-in on my spare time as volunteer, so if family/$work/summer/other OSS project issues/drawing/etc. happen, then I will come back to fix that once I have time to work on this again. But let's hope I can revert, test, and prepare a release soon.</p>

<p>Cheers</p>
</pre>
</details>

</ul>
</details>

<p>Hi, thanks for the detailed report. I created a new pipeline with your example, then played a bit with the returned Groovy (see screenshot).</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/62948/image-2024-07-06-10-43-20-297.png" style="border: 0px solid black" /></span></p>

<p>And I believe what's happening here is that <a href="https://issue-redirect.jenkins.io/issue/71909?filter=14783" class="external-link" rel="nofollow">https://issue-redirect.jenkins.io/issue/71909?filter=14783</a> is affecting your script now too.</p>

<p>Previously, the parameters were loaded in order. So your document.getElementById would succeed as the previous parameter ("I", with the "CB_BUILDS") would have been rendered first.</p>

<p>This issue linked above was a tentative to optimize the rendering of parameters, by rendering them as fast as possible. Which at least sounds like a good idea, but seems like it's creating issues for several users.</p>

<p>I will not work on your issue for now, and instead will try to use this remaining spare time that I have today/tomorrow to revert that change.</p>

<p>Please, note that if this takes a bit, it's because I work on the plug-in on my spare time as volunteer, so if family/$work/summer/other OSS project issues/drawing/etc. happen, then I will come back to fix that once I have time to work on this again. But let's hope I can revert, test, and prepare a release soon.</p>

<p>Cheers</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/hossted.png" title="hossted's avatar" /> hossted:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=449891&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-449891">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Greetings!</p>

<p>We are also <b>a community member</b> experiencing the similar issue.</p>

<p>We hired professional expert(s) to work on this matter since our business case demanded so.
We submitted the dedicated issue describing the problem(s) in detail here:
<a href="https://issues.jenkins.io/browse/JENKINS-73935" class="external-link" rel="nofollow">JENKINS-73935</a></p>

<p>We hope that the <b>plugin maintainer(s) and / or author(s) will be able to fix the root cause of this issue</b> in the near future.</p>

<p>In the meantime, we created a separate repository project which contains a workaround for this issue.</p>

<p>For more information, you can check it out <a href="https://github.com/hossted-prs/jenkins-ac-plugin-cascading-workaround/" class="external-link" target="_blank" rel="nofollow noopener">here (on our GitHub)</a>.</p>

<p>Please read in detail if you are interested.
Also, we provided a patched plugin bundles containing the fix <a href="https://github.com/hossted-prs/jenkins-ac-plugin-cascading-workaround/releases/" class="external-link" target="_blank" rel="nofollow noopener">here (on our GitHub)</a>.</p>

<p>But due to security concerns, please, use with caution.</p>

<p>If you are interested in using the fix before the official fix from the maintainer(s), it would be preferred that you rebuild it yourself instead of using our pre-built bundle we are distributing.</p>

<p> </p>

<p>Best regards,
<a href="https://github.com/hossted-prs" class="external-link" target="_blank" rel="nofollow noopener">Hossted</a></p>
</pre>
</details>

</ul>
</details>

<p>Greetings!</p>

<p>We are also <b>a community member</b> experiencing the similar issue.</p>

<p>We hired professional expert(s) to work on this matter since our business case demanded so.<br/>
We submitted the dedicated issue describing the problem(s) in detail here:<br/>
<a href="https://issue-redirect.jenkins.io/issue/73935" class="external-link" rel="nofollow">JENKINS-73935</a></p>

<p>We hope that the <b>plugin maintainer(s) and / or author(s) will be able to fix the root cause of this issue</b> in the near future.</p>

<p>In the meantime, we created a separate repository project which contains a workaround for this issue.</p>

<p>For more information, you can check it out <a href="https://github.com/hossted-prs/jenkins-ac-plugin-cascading-workaround/" class="external-link" target="_blank" rel="nofollow noopener">here (on our GitHub)</a>.</p>

<p>Please read in detail if you are interested.<br/>
Also, we provided a patched plugin bundles containing the fix <a href="https://github.com/hossted-prs/jenkins-ac-plugin-cascading-workaround/releases/" class="external-link" target="_blank" rel="nofollow noopener">here (on our GitHub)</a>.</p>

<p>But due to security concerns, please, use with caution.</p>

<p>If you are interested in using the fix before the official fix from the maintainer(s), it would be preferred that you rebuild it yourself instead of using our pre-built bundle we are distributing.</p>

<p> </p>

<p>Best regards,<br/>
<a href="https://github.com/hossted-prs" class="external-link" target="_blank" rel="nofollow noopener">Hossted</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=453304&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453304">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Injecting JavaScript like that won't work as there was a severe security issue, and now all JS is escaped in the plug-in (otherwise, the plug-in is marked as not safe – updates prevented, etc.).</p>

<p>But I found one issue preventing this issue from being closed. With the example attached, after modifying a bit, I realized the cascade parameters are not referencing anything. This is something we don't have tests for, and that are causing JS errors, which could introduce new bugs if this version was released as-is.</p>

<p>I'll write a test to fix, and modify the script so it doesn't use JS. Once the PR with the test is merged this will be closed. I will see if I can find the issue where the JS started being escaped, and I think we have an older issue about someone that needed/wanted it. I will close this as duplicated of that.</p>

<p>Cheers</p>
</pre>
</details>

</ul>
</details>

<p>Injecting JavaScript like that won't work as there was a severe security issue, and now all JS is escaped in the plug-in (otherwise, the plug-in is marked as not safe – updates prevented, etc.).</p>

<p>But I found one issue preventing this issue from being closed. With the example attached, after modifying a bit, I realized the cascade parameters are not referencing anything. This is something we don't have tests for, and that are causing JS errors, which could introduce new bugs if this version was released as-is.</p>

<p>I'll write a test to fix, and modify the script so it doesn't use JS. Once the PR with the test is merged this will be closed. I will see if I can find the issue where the JS started being escaped, and I think we have an older issue about someone that needed/wanted it. I will close this as duplicated of that.</p>

<p>Cheers</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=453654&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453654">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Exception logged: "Uncaught TypeError: dataHolder.dataset.referencedParameters is undefined"</p>

<p> </p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/64211/64211_image-2025-04-06-19-38-17-470.png" height="446" width="494" style="border: 0px solid black" /></span></p>
</pre>
</details>

</ul>
</details>

<p>Exception logged: "Uncaught TypeError: dataHolder.dataset.referencedParameters is undefined"</p>

<p> </p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/64211/image-2025-04-06-19-38-17-470.png" height="446" width="494" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=453655&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453655">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>With the change in the code, now it will log:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/64212/64212_image-2025-04-06-19-40-19-325.png" height="470" width="518" style="border: 0px solid black" /></span></p>

<p>Writing the test for that now.</p>
</pre>
</details>

</ul>
</details>

<p>With the change in the code, now it will log:</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/64212/image-2025-04-06-19-40-19-325.png" height="470" width="518" style="border: 0px solid black" /></span></p>

<p>Writing the test for that now.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=453657&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453657">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Done! <a href="https://github.com/jenkinsci/active-choices-plugin/pull/467" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/active-choices-plugin/pull/467</a></p>
</pre>
</details>

</ul>
</details>

<p>Done! <a href="https://github.com/jenkinsci/active-choices-plugin/pull/467" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/active-choices-plugin/pull/467</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-73239?focusedId=453658&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453658">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Closing as Won't Fix as Jenkins security prevents the plugin from loading and executing third party JS scripts. If you would like this feature, open an issue against Jenkins Core requesting a secure way for plug-ins to do so.</p>
</pre>
</details>

</ul>
</details>

<p>Closing as Won't Fix as Jenkins security prevents the plugin from loading and executing third party JS scripts. If you would like this feature, open an issue against Jenkins Core requesting a secure way for plug-ins to do so.</p>
